### PR TITLE
Some code quality fixes

### PR DESCRIPTION
**Pull request by oscargus**
#### Brief description of what is fixed or changed
I just learnt about LGTM: https://lgtm.com/projects/g/sympy/sympy/overview/

Hence, I decided to fix some issues pointed out there.

I realized that the `integrals.rubi` part is not currently tested, but I still fixed an issue and added a test for that issue there. Quite a lot of issues pointed out about that code otherwise.

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by oscargus** [comment]
By the way: who is in charge of the LGTM connection? @asmeurer @smichr 

It seems like it is currently timing out and maybe there is nothing we can do about that, but it would be nice to at least try to increase the time.

File: sympy/integrals/prde.py [hunk 1]
```diff
@@ -773,15 +773,15 @@ def limited_integrate_reduce(fa, fd, G, DE):
...
+    b = derivation(hn, DE)
     N = 0
 
     # These are the cases where we know that S1irr = Sirr, but there could be
     # others, and this algorithm will need to be extended to handle them.
     if DE.case in ['base', 'primitive', 'exp', 'tan']:
         hs = reduce(lambda i, j: i.lcm(j), (ds,) + Es)  # lcm(ds, es1, ..., esm)
         a = hn*hs
-        b = -derivation(hn, DE) - (hn*derivation(hs, DE)).quo(hs)
+        b = -b - (hn*b).quo(hs)
```
Comment (smichr) [inline sympy/integrals/prde.py]: this looks suspiciously wrong since it identifies derivation(hn, DE) and derivation(**hs**, DE) as b

Comment (oscargus) [reply sympy/integrals/prde.py]: Definitely! Well spotted! Thanks.

File: sympy/integrals/prde.py [hunk 2]
```diff
@@ -773,15 +773,15 @@ def limited_integrate_reduce(fa, fd, G, DE):
     c = reduce(lambda i, j: i.lcm(j), (dn,) + En)  # lcm(dn, en1, ..., enm)
     hn = c.gcd(c.diff(DE.t))
     a = hn
-    b = -derivation(hn, DE)
+    b = derivation(hn, DE)
```
Comment (jksuom) [inline sympy/integrals/prde.py]: I would have preserved the minus sign here. It comes from the pseudocode in the book.

**Comment by asmeurer** [comment]
I had no idea lgtm was running on SymPy. As far as I can tell it just does it on its own, or someone told it to start following SymPy and now it does.

File: sympy/integrals/prde.py (hunk 2 above)
Comment (oscargus) [reply sympy/integrals/prde.py]: Sure! No problems. I only removed it because of the incorrect reuse.

**Comment by asmeurer** [comment]
It does seem to be an interesting service. I like that it focuses only on code correctness issues, not style issues. It would be nice if they allowed suppressing alerts without modifying the source code.

**Comment by oscargus** [comment]
@asmeurer OK! It looks like someone has some sort of control over it. It shows that I am following sympy, but I cannot access anything (and I guess that someone can).

Your contributions show up quite nicely in the graph! https://lgtm.com/projects/g/sympy/sympy/contributors:python

I also like the idea. It shows some more issues compared to pylint etc and seems to analyze the code more thoroughly (even though there are some false positives). Apart from the fact that it currently times out (after four hours...) it would be a nice thing to see if a PR adds or removes issues here at Github.

File: sympy/integrals/prde.py [hunk 3]
```diff
@@ -781,7 +781,7 @@ def limited_integrate_reduce(fa, fd, G, DE):
     if DE.case in ['base', 'primitive', 'exp', 'tan']:
         hs = reduce(lambda i, j: i.lcm(j), (ds,) + Es)  # lcm(ds, es1, ..., esm)
         a = hn*hs
-        b = -derivation(hn, DE) - (hn*derivation(hs, DE)).quo(hs)
+        b -= (hn*derivation(hs, DE)).quo(hs)
```
Comment (oscargus) [inline sympy/integrals/prde.py]: Thanks @smichr !

**Comment by asmeurer** [comment]
Some background https://github.com/sympy/sympy/pull/12816

**Comment by asmeurer** [comment]
@rneatherway some lgtm questions here:

- Is there a person who "controls" the sympy/sympy lgtm page? As far as I can tell, lgtm just looks at public repos and doesn't require access unless you set up integrations. Is this correct?

- Is there a way to ignore warnings from within the lgtm interface, rather than by adding a comment to the source code? I think it would be useful to have the info be stored on lgtm rather than in the repo. But maybe it's not a trivial feature because lines of code can move around, so you'd have to have some `git blame`-like heuristics.

**Comment by rneatherway** [comment]
Hi @oscargus , @asmeurer,

> Is there a person who "controls" the sympy/sympy lgtm page? As far as I can tell, lgtm just looks at public repos and doesn't require access unless you set up integrations. Is this correct?

Yes, we try to analyse as many significant public repos as we can on LGTM so that the results are there ready for people to look at. So in that sense nobody 'controls' the page, that's correct. To enable PR integration you need to be an admin on the repository.

> Is there a way to ignore warnings from within the lgtm interface, rather than by adding a comment to the source code? I think it would be useful to have the info be stored on lgtm rather than in the repo. But maybe it's not a trivial feature because lines of code can move around, so you'd have to have some git blame-like heuristics.

Right, currently there isn't a way to do that. The neat thing about the comment in the source code is that it follows the alert around naturally. You can [disable certain queries](https://lgtm.com/help/lgtm/showing-hiding-query-results) if you like, for example if they aren't relevant or are too noisy for your code base.

The issue of tracking alerts when they move around is a very interesting one, and we do have some fairly sophisticated methods of doing that, which are used to populate the [My alerts](https://lgtm.com/active_alerts?mode=tree&showSuggestions=false) page. This shows just the alerts that you have introduced, plus optionally other alerts near to code that you have modified. Whether we should also use this feature to allow hiding alerts in LGTM is something we haven't decided on yet. For example, should they only be hidden for you, or for everyone? Keeping them in the repository offers a good audit trail and ensures that not just everyone can add them.

> Apart from the fact that it currently times out (after four hours...) it would be a nice thing to see if a PR adds or removes issues here at Github.

I've checked in with our Python team and this is a performance regression in the analysis that we believe should be resolved when we deploy the next version of the analysis engine. I'll take a note to let you know when it is back up and running and you can decide if you'd like to try PR analysis then.

**Comment by oscargus** [comment]
Thanks for the clarification @rneatherway ! Looking forward to when the new version arrives as I really like this type of tools...

**Comment by rneatherway** [comment]
Hi all, sorry for the delay in getting back to you. I'm happy to let you know that LGTM is now analysing sympy successfully again. You can see the latest list of alerts from yesterday at https://lgtm.com/projects/g/sympy/sympy/alerts/?mode=list. If you have any feedback on the results or questions feel free to raise them at https://discuss.lgtm.com/ and I or one of my colleagues will try to help.

**Pull request merged**